File: packages/nightgauge-sdk/src/events/phaseInference.ts
/**
 * Phase Inference — Deterministic phase progress from observable tool activity
 *
 * Some skills do not reliably emit `<!-- phase:start ... -->` markers. The
 * `feature-dev` stage in particular is edit-heavy (Read/Edit/Write dominate its
 * tool calls; Bash is rare), so the model routinely skips the standalone
 * `printf` phase-marker commands scattered through its SKILL.md. The result is
 * that the pipeline tree shows no phase progress for Feature Development even
 * though planning and validation render fine (Issue #3760).
 *
 * Hardening the marker parser or delivery path cannot fix this — a marker that
 * is never emitted cannot be delivered. Instead, this module infers phase
 * progress deterministically from the tool calls the agent actually makes,
 * which the orchestrator already observes in the stream. The inferred markers
 * are fed through the exact same `onPhaseStart` channel that real markers use,
 * so they render identically.
 *
 * Design guarantees:
 * - **Monotonic**: the inferred phase cursor only ever advances, never regresses
 *   (a Read during the implementation phase will not snap back to an early
 *   context phase).
 * - **Real markers win**: when a skill *does* emit a genuine marker, call
 *   `observeRealMarker(index)` so inference syncs its cursor forward and never
 *   duplicates or contradicts it.
 * - **Opt-in per stage**: only stages with a rule table (currently `feature-dev`)
 *   produce inferred markers. Stages that already emit reliably are untouched —
 *   their inference is a no-op.
 *
 * @see Issue #3760 - Feature Development stage shows no phases in pipeline tree
 */

import { PHASE_REGISTRY, type ExecutionStage, type ParsedPhaseMarker } from "./phaseRegistry.js";

/**
 * A rule that maps an observed tool call to a target phase index within a stage.
 */
interface PhaseInferenceRule {
  /** Target 0-based phase index within the stage (must exist in PHASE_REGISTRY). */
  index: number;
  /** Predicate over an observed tool call. Returns true when this phase is reached. */
  match: (toolName: string, toolInput: unknown) => boolean;
}

/** Safely read a string field from an unknown tool-input object. */
function inputString(input: unknown, key: string): string {
  if (input && typeof input === "object") {
    const v = (input as Record<string, unknown>)[key];
    if (typeof v === "string") return v;
  }
  return "";
}

const EDIT_TOOLS = new Set(["Edit", "Write", "MultiEdit", "NotebookEdit"]);
const READ_TOOLS = new Set(["Read", "Grep", "Glob"]);

/** Paths that represent pipeline bookkeeping rather than feature source code. */
function isPipelineArtifactPath(path: string): boolean {
  return /(^|\/)\.nightgauge\//.test(path);
}

/** The dev-context handoff file written near the end of feature-dev. */
function isDevContextPath(path: string): boolean {
  return /(^|\/)dev-\d+\.json$/.test(path) || /\.nightgauge\/pipeline\/dev-/.test(path);
}

/** Bash commands that signal the testing / build-verification phase. */
function isTestOrBuildCommand(cmd: string): boolean {
  return /\b(vitest|jest|go\s+test|go\s+build|npm\s+(run\s+)?(-w\s+\S+\s+)?(test|build)|pytest|cargo\s+test)\b/.test(
    cmd
  );
}

/** Bash commands that signal the project-status sync phase. */
function isStatusSyncCommand(cmd: string): boolean {
  return /\b(move-status|gh\s+project)\b/.test(cmd);
}

/** The plan file written mid-way through feature-planning (`.nightgauge/plans/{N}-*.md`). */
function isPlanFilePath(path: string): boolean {
  return /(^|\/)\.nightgauge\/plans\/.+\.md$/.test(path);
}

/** The planning-context handoff file written near the end of feature-planning. */
function isPlanningContextPath(path: string): boolean {
  return /(^|\/)planning-\d+\.json$/.test(path) || /\.nightgauge\/pipeline\/planning-/.test(path);
}

/**
 * Per-stage inference rules. Only stages that do NOT reliably self-report phase
 * markers need entries here. Rules are evaluated against each tool call; when
 * several match, the highest index strictly greater than the current cursor
 * wins (monotonic advancement). Indices must match PHASE_REGISTRY ordering.
 */
const STAGE_RULES: Partial<Record<ExecutionStage, PhaseInferenceRule[]>> = {
  // feature-dev (18 phases). Maps the durable, observable waypoints of an
  // edit-heavy implementation run to their canonical phase indices.
  "feature-dev": [
    // Context loading (covers phases 0-7, which run in seconds at the start).
    { index: 1, match: (name) => READ_TOOLS.has(name) },
    // First edit to real source code → implementation. Excludes pipeline
    // bookkeeping and the dev-context handoff file so those don't mis-fire here.
    {
      index: 8,
      match: (name, input) => {
        if (!EDIT_TOOLS.has(name)) return false;
        const path = inputString(input, "file_path") || inputString(input, "notebook_path");
        return !!path && !isPipelineArtifactPath(path) && !isDevContextPath(path);
      },
    },
    // Running the test/build suite → testing.
    {
      index: 9,
      match: (name, input) =>
        name === "Bash" && isTestOrBuildCommand(inputString(input, "command")),
    },
    // Writing the downstream dev context handoff → write-dev-context.
    {
      index: 14,
      match: (name, input) =>
        EDIT_TOOLS.has(name) && isDevContextPath(inputString(input, "file_path")),
    },
    // Syncing project board status → sync-project-status.
    {
      index: 15,
      match: (name, input) => name === "Bash" && isStatusSyncCommand(inputString(input, "command")),
    },
  ],
  // feature-planning (14 phases). Read-dominated: the agent spends most of its
  // time reading docs/standards/source before writing the plan and the
  // planning-context handoff. When planning runs on an edit-heavy model (e.g.
  // sonnet rather than haiku) it skips the standalone printf markers, so infer
  // from the durable observable waypoints. Real markers still win (#3771).
  "feature-planning": [
    // Reading docs/standards/source → documentation-analysis, where planning
    // spends the bulk of its time. Covers the fast early phases 0-6.
    { index: 6, match: (name) => READ_TOOLS.has(name) },
    // Writing the plan file (.nightgauge/plans/{N}-*.md) → produce-plan.
    {
      index: 9,
      match: (name, input) =>
        EDIT_TOOLS.has(name) && isPlanFilePath(inputString(input, "file_path")),
    },
    // Writing the planning-context handoff → write-planning-context.
    {
      index: 10,
      match: (name, input) =>
        EDIT_TOOLS.has(name) && isPlanningContextPath(inputString(input, "file_path")),
    },
  ],
};

/**
 * A stateful phase inferrer for a single stage run.
 */
export interface PhaseInference {
  /** True when this stage has inference rules (i.e. inference is active). */
  readonly enabled: boolean;
  /**
   * Emit the stage's initial phase. Call once when the stage starts so the tree
   * shows a live phase immediately. Returns the marker to emit, or null.
   */
  start(): ParsedPhaseMarker | null;
  /**
   * Observe a tool call. Returns a marker to emit when it advances the phase
   * past the current cursor, otherwise null.
   */
  observeToolUse(toolName: string, toolInput: unknown): ParsedPhaseMarker | null;
  /**
   * Sync the cursor when the skill emitted a genuine marker, so inferred
   * markers never regress or duplicate a real one. No marker is returned —
   * the real marker is delivered through the normal path.
   */
  observeRealMarker(index: number): void;
}

/**
 * Build a parsed marker for a given stage/index from the canonical registry.
 * Returns null if the index is out of range for the stage.
 */
function markerFor(stage: ExecutionStage, index: number): ParsedPhaseMarker | null {
  const phases = PHASE_REGISTRY[stage];
  if (!phases || index < 0 || index >= phases.length) return null;
  return { name: phases[index].name, index, total: phases.length, stage };
}

/**
 * Create a phase inferrer for a stage. For stages without rules, all methods are
 * no-ops (`enabled === false`), so reliably-self-reporting stages are unaffected.
 *
 * @param stage - The pipeline stage name
 * @returns A stateful {@link PhaseInference}
 */
export function createPhaseInference(stage: string): PhaseInference {
  const rules = STAGE_RULES[stage as ExecutionStage];
  const execStage = stage as ExecutionStage;
  const enabled = !!rules && !!PHASE_REGISTRY[execStage];

  // Highest phase index emitted/observed so far. -1 means nothing yet.
  let cursor = -1;

  function advanceTo(index: number): ParsedPhaseMarker | null {
    if (index <= cursor) return null;
    const marker = markerFor(execStage, index);
    if (!marker) return null;
    cursor = index;
    return marker;
  }

  return {
    enabled,

    start(): ParsedPhaseMarker | null {
      if (!enabled) return null;
      return advanceTo(0);
    },

    observeToolUse(toolName: string, toolInput: unknown): ParsedPhaseMarker | null {
      if (!enabled || !rules) return null;
      let best = -1;
      for (const rule of rules) {
        if (rule.index > best && rule.index > cursor && rule.match(toolName, toolInput)) {
          best = rule.index;
        }
      }
      if (best === -1) return null;
      return advanceTo(best);
    },

    observeRealMarker(index: number): void {
      if (!enabled) return;
      if (index > cursor) cursor = index;
    },
  };
}

File: internal/execution/phase_inference.go
package execution

import (
	"encoding/json"
	"regexp"

	"github.com/nightgauge/nightgauge/internal/state"
)

// Phase inference — deterministic phase progress from observable tool activity.
//
// Some skills do not reliably emit `<!-- phase:start ... -->` markers. The
// feature-dev stage in particular is edit-heavy (Read/Edit/Write dominate its
// tool calls; Bash is rare), so the model routinely skips the standalone printf
// phase-marker commands in its SKILL.md. The result is that the pipeline tree
// shows no phase progress for Feature Development even though planning and
// validation render fine (Issue #3760).
//
// Hardening the marker parser cannot fix this — a marker that is never emitted
// cannot be parsed. Instead this infers phase progress from the tool calls the
// agent actually makes, which the execution manager already scans, and feeds
// the inferred markers through the same PhaseEventFn channel as real markers.
//
// This mirrors packages/nightgauge-sdk/src/events/phaseInference.ts so the
// Go (auto/CLI) and TypeScript (VSCode/IPC) execution paths behave identically.

// stagePhaseTables maps a stage to its ordered phase names, read from
// internal/state's PhaseRegistry — the one Go-side declaration of these
// tables (#1885). Only stages that do NOT reliably self-report phase markers
// need an entry; others are no-ops.
var stagePhaseTables = map[string][]string{
	"feature-dev":      state.RegistryPhaseNames(state.StageFeatureDev),
	"feature-planning": state.RegistryPhaseNames(state.StageFeaturePlanning),
	"feature-validate": state.RegistryPhaseNames(state.StageFeatureValidate),
}

var (
	editToolRe        = regexp.MustCompile(`^(Edit|Write|MultiEdit|NotebookEdit)$`)
	readToolRe        = regexp.MustCompile(`^(Read|Grep|Glob)$`)
	pipelinePathRe    = regexp.MustCompile(`(^|/)\.nightgauge/`)
	devContextRe      = regexp.MustCompile(`(^|/)dev-\d+\.json$|\.nightgauge/pipeline/dev-`)
	testBuildRe       = regexp.MustCompile(`\b(vitest|jest|go\s+test|go\s+build|npm\s+(run\s+)?(-w\s+\S+\s+)?(test|build)|pytest|cargo\s+test)\b`)
	statusSyncRe      = regexp.MustCompile(`\b(move-status|gh\s+project)\b`)
	planFileRe        = regexp.MustCompile(`(^|/)\.nightgauge/plans/.+\.md$`)
	planningContextRe = regexp.MustCompile(`(^|/)planning-\d+\.json$|\.nightgauge/pipeline/planning-`)
	validateContextRe = regexp.MustCompile(`(^|/)validate-\d+\.json$|\.nightgauge/pipeline/validate-`)
	gitPushRe         = regexp.MustCompile(`\bgit\s+push\b`)
)

// inferenceRule maps an observed tool call to a target phase index.
type inferenceRule struct {
	index int
	match func(toolName string, input map[string]any) bool
}

func inputStr(input map[string]any, key string) string {
	if input == nil {
		return ""
	}
	if v, ok := input[key].(string); ok {
		return v
	}
	return ""
}

// stageRules returns the ordered inference rules for a stage, or nil.
func stageRules(stage string) []inferenceRule {
	if _, ok := stagePhaseTables[stage]; !ok {
		return nil
	}
	switch stage {
	case "feature-dev":
		return []inferenceRule{
			{index: 1, match: func(name string, _ map[string]any) bool { return readToolRe.MatchString(name) }},
			{index: 8, match: func(name string, input map[string]any) bool {
				if !editToolRe.MatchString(name) {
					return false
				}
				path := inputStr(input, "file_path")
				if path == "" {
					path = inputStr(input, "notebook_path")
				}
				return path != "" && !pipelinePathRe.MatchString(path) && !devContextRe.MatchString(path)
			}},
			{index: 9, match: func(name string, input map[string]any) bool {
				return name == "Bash" && testBuildRe.MatchString(inputStr(input, "command"))
			}},
			{index: 14, match: func(name string, input map[string]any) bool {
				return editToolRe.MatchString(name) && devContextRe.MatchString(inputStr(input, "file_path"))
			}},
			{index: 15, match: func(name string, input map[string]any) bool {
				return name == "Bash" && statusSyncRe.MatchString(inputStr(input, "command"))
			}},
		}
	case "feature-planning":
		return []inferenceRule{
			// Reading docs/standards/source → documentation-analysis, where
			// planning spends the bulk of its time. Covers early phases 0-6.
			{index: 6, match: func(name string, _ map[string]any) bool { return readToolRe.MatchString(name) }},
			// Writing the plan file (.nightgauge/plans/{N}-*.md) → produce-plan.
			{index: 9, match: func(name string, input map[string]any) bool {
				return editToolRe.MatchString(name) && planFileRe.MatchString(inputStr(input, "file_path"))
			}},
			// Writing the planning-context handoff → write-planning-context.
			{index: 10, match: func(name string, input map[string]any) bool {
				return editToolRe.MatchString(name) && planningContextRe.MatchString(inputStr(input, "file_path"))
			}},
		}
	case "feature-validate":
		return []inferenceRule{
			// Any read → read-dev-context. The stage opens by reading the dev
			// handoff and the files it names, so this is the earliest honest
			// waypoint and covers phases 0-1.
			{index: 1, match: func(name string, _ map[string]any) bool { return readToolRe.MatchString(name) }},
			// A build/test command → run-tests, where the stage spends nearly
			// all of its wall-clock. Deliberately the same testBuildRe
			// feature-dev uses: the two stages run the same commands, and a
			// second pattern for one vocabulary is a drift source.
			{index: 10, match: func(name string, input map[string]any) bool {
				return name == "Bash" && testBuildRe.MatchString(inputStr(input, "command"))
			}},
			// Pushing the branch → commit-and-push. Anchored on `git push`
			// rather than any git call, because the stage runs `git status` and
			// `git diff` throughout and neither is this waypoint.
			{index: 18, match: func(name string, input map[string]any) bool {
				return name == "Bash" && gitPushRe.MatchString(inputStr(input, "command"))
			}},
			// Writing the validate-context handoff → write-validate-context.
			{index: 19, match: func(name string, input map[string]any) bool {
				return editToolRe.MatchString(name) && validateContextRe.MatchString(inputStr(input, "file_path"))
			}},
			// Board status sync → sync-project-status, the stage's last
			// observable act before it narrates.
			{index: 20, match: func(name string, input map[string]any) bool {
				return name == "Bash" && statusSyncRe.MatchString(inputStr(input, "command"))
			}},
		}
	default:
		return nil
	}
}

// PhaseInferer infers phase progress for a single stage run. Monotonic: the
// cursor only ever advances. Real markers take precedence via ObserveRealMarker.
type PhaseInferer struct {
	stage   string
	phases  []string
	rules   []inferenceRule
	enabled bool
	cursor  int // highest phase index emitted/observed; -1 = none yet
}

// NewPhaseInferer builds an inferer for a stage. For stages without rules it is
// disabled and all methods are no-ops, leaving self-reporting stages untouched.
func NewPhaseInferer(stage string) *PhaseInferer {
	phases := stagePhaseTables[stage]
	rules := stageRules(stage)
	return &PhaseInferer{
		stage:  stage,
		phases: phases,
		rules:  rules,
		// A phase table alone is enough (#1924). Gap-fill derives progress from
		// ORDERING, so it works for a stage that self-reports markers and has no
		// inference rules at all; rules only add extra landmarks to fill between.
		// Requiring both is what kept gap-fill off any stage the rule table had
		// not been taught yet.
		enabled: len(phases) > 0,
		cursor:  -1,
	}
}

func (p *PhaseInferer) markerFor(index int) (*PhaseMarker, bool) {
	if index < 0 || index >= len(p.phases) {
		return nil, false
	}
	return &PhaseMarker{
		Name:  p.phases[index],
		Index: index,
		Total: len(p.phases),
		Stage: p.stage,
	}, true
}

// advanceTo moves the cursor to index and returns the marker for it, plus the
// markers for every phase the cursor jumped OVER (#1924).
//
// The jumped-over phases are the whole point: a run that reports 1 and then 9
// was in 2..8 and said nothing, and before this they vanished until the
// end-of-stage back-fill. They are returned separately from the advance because
// they are a weaker claim — see RuntimeState.PassPhase.
func (p *PhaseInferer) advanceTo(index int) (*PhaseMarker, []PhaseMarker, bool) {
	if index <= p.cursor {
		return nil, nil, false
	}
	m, ok := p.markerFor(index)
	if !ok {
		return nil, nil, false
	}
	passed := p.gapBelow(index)
	p.cursor = index
	return m, passed, true
}

// gapBelow returns markers for the unreported phases strictly between the
// cursor and index. Callers hold no lock; PhaseInferer is single-goroutine,
// driven by the one output-scanning loop in manager.go.
func (p *PhaseInferer) gapBelow(index int) []PhaseMarker {
	var passed []PhaseMarker
	for i := p.cursor + 1; i < index; i++ {
		if m, ok := p.markerFor(i); ok {
			passed = append(passed, *m)
		}
	}
	return passed
}

// Start emits the stage's first phase. Call once when output begins.
func (p *PhaseInferer) Start() (*PhaseMarker, bool) {
	if !p.enabled {
		return nil, false
	}
	m, _, ok := p.advanceTo(0)
	return m, ok
}

// ObserveToolUse returns a marker when the tool call advances the phase.
func (p *PhaseInferer) ObserveToolUse(toolName string, input map[string]any) (*PhaseMarker, []PhaseMarker, bool) {
	if !p.enabled {
		return nil, nil, false
	}
	best := -1
	for _, r := range p.rules {
		if r.index > best && r.index > p.cursor && r.match(toolName, input) {
			best = r.index
		}
	}
	if best == -1 {
		return nil, nil, false
	}
	return p.advanceTo(best)
}

// ObserveRealMarker syncs the cursor forward when a genuine marker was emitted,
// so inferred markers never regress or duplicate a real one.
// A real marker also closes a gap: a skill that printf's 0 then 6 was in 1..5
// and never said so, so the jumped-over phases are returned here too (#1924).
func (p *PhaseInferer) ObserveRealMarker(index int) []PhaseMarker {
	if !p.enabled || index <= p.cursor {
		return nil
	}
	passed := p.gapBelow(index)
	p.cursor = index
	return passed
}

// toolUse is a single tool call extracted from an assistant message.
type toolUse struct {
	Name  string
	Input map[string]any
}

// openCodeToolNameToClaudeName maps an OpenCode tool_use event's part.tool
// (opencode's own lowercase tool id, ADR-022 § 1 / opencode_usage.go) to the
// Claude tool name stageRules matches on, so the one rule table drives phase
// inference on both the Claude assistant-message shape and OpenCode's. write
// and apply_patch join edit: opencode 1.18.30 has no permission or usage
// distinction between them (openCodeToolRejectionPermission), and neither
// does an edit-heavy stage's phase progress. A tool this table has no entry
// for (task, webfetch, websearch, todowrite, skill, lsp, ...) matches no
// phase rule, same as an unrecognized Claude tool name would.
var openCodeToolNameToClaudeName = map[string]string{
	"bash":        "Bash",
	"edit":        "Edit",
	"write":       "Edit",
	"apply_patch": "Edit",
	"read":        "Read",
	"glob":        "Glob",
	"grep":        "Grep",
}

// extractOpenCodeToolUse reads one OpenCode `tool_use` event (opencode
// 1.18.30's `run --format json`, ADR-022): `{"type":"tool_use","part":
// {"type":"tool","tool":"edit","state":{"input":{"filePath":...}}}}`. OpenCode
// emits this only once a call has completed or errored, never on start
// (#1635's plugin-handshake comment records the same observation), so a
// phase this advances to was actually reached, not merely attempted.
//
// filePath, OpenCode's input key for edit/write/apply_patch, is copied to
// file_path so the SAME rules that read Claude's Edit/Write input
// (inputStr(input, "file_path")) match here without a second rule table.
func extractOpenCodeToolUse(line string) (toolUse, bool) {
	var env struct {
		Type string `json:"type"`
		Part struct {
			Type  string `json:"type"`
			Tool  string `json:"tool"`
			State struct {
				Input map[string]any `json:"input"`
			} `json:"state"`
		} `json:"part"`
	}
	if err := json.Unmarshal([]byte(line), &env); err != nil {
		return toolUse{}, false
	}
	if env.Type != "tool_use" || env.Part.Type != "tool" {
		return toolUse{}, false
	}
	name, ok := openCodeToolNameToClaudeName[env.Part.Tool]
	if !ok {
		return toolUse{}, false
	}
	input := env.Part.State.Input
	if fp, ok := input["filePath"].(string); ok {
		input["file_path"] = fp
	}
	return toolUse{Name: name, Input: input}, true
}

// extractToolUses parses one line of stage stdout and returns its tool_use
// blocks: an assistant stream-json line (the CLI's shape, tool calls inside
// complete `assistant` messages) or an OpenCode `tool_use` event
// (extractOpenCodeToolUse). Returns nil for a line that is neither.
func extractToolUses(line string) []toolUse {
	var env struct {
		Type    string `json:"type"`
		Message struct {
			Content []struct {
				Type  string         `json:"type"`
				Name  string         `json:"name"`
				Input map[string]any `json:"input"`
			} `json:"content"`
		} `json:"message"`
	}
	if err := json.Unmarshal([]byte(line), &env); err == nil && env.Type == "assistant" {
		var out []toolUse
		for _, b := range env.Message.Content {
			if b.Type == "tool_use" && b.Name != "" {
				out = append(out, toolUse{Name: b.Name, Input: b.Input})
			}
		}
		return out
	}
	if tu, ok := extractOpenCodeToolUse(line); ok {
		return []toolUse{tu}
	}
	return nil
}
